### DISCUSSION TITLE
User messages

### DISCUSSION THREAD

**Discussion by span** [ActiveLogin/ActiveLogin.Authentication#392] [General]
Hi!

In version 6 it seems the ability to override user messages was changed or removed. Previously all messages were in the resource files. I realise BankID have their recommended messages, but some of them surely do have some readability issues and needs overriding.

How, if possible, can the user messages be changed and localised in v6?

**Comment by PeterOrneholm** [comment]
Hi @span!

Good question! This is something you can still override. When we moved most of our logic into the Core package, we also wanted to move the texts. Moving the `.resx` files turned out harder than we thought, so we decided to split them. The `.resx` files still in the project are there for texts in the web UI. Like buttons etc. The User messages from BankID are instead implemented in this clas:
https://github.com/ActiveLogin/ActiveLogin.Authentication/blob/main/src/ActiveLogin.Authentication.BankId.Core/UserMessage/BankIdUserMessageStringLocalizer.cs

I would inherit that class and override the specific texts you want to change, and for the rest just call the base implementation. By doing so you would get the texts we add in the future.

If you want to place your texts in a `.resx`-file, the implementation is super simple:
https://github.com/ActiveLogin/ActiveLogin.Authentication/blob/v5.0.0/src/ActiveLogin.Authentication.BankId.AspNetCore/UserMessage/BankIdUserMessageStringLocalizer.cs

**Comment by span** [comment]
Thanks. I'll have another look. I am implementing `IBankIdUserMessageLocalizer` but my resources from resx files are not used. Will investigate further!

**Reply by span** [reply]
Ok, so I had another look.

The reason it stopped working for us was that we were now adding the `IBankIdUserMessageLocalizer` before calling  `AddBankId`. This caused our implementation to be overwritten by the default.

I was not expecting this since I thought the library was using the `TryAddTransient` extensions to add to DI. Turns out I was wrong.

As a side note I would suggest using `TryAdd` to avoid unintentional overrides in the service configuration, or perhaps this has already been considered?

Thanks again for quick response :D

**Reply by PeterOrneholm** [reply]
Great! Hmm, I think we had to change that (for some reason). Our recomendation is, at least from now on then, to register any cusotm implementation after .AddBankId().
